File: backend-dotnet/SahayaSetu/Controllers/NGOController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SahayaSetu.DTO.NGO;
using SahayaSetu.DTO.Request;
using SahayaSetu.DTO; // Added back
using SahayaSetu.Services;

namespace SahayaSetu.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class NGOController : ControllerBase
    {
        private readonly INgoService _ngoService;

        public NGOController(INgoService ngoService)
        {
            _ngoService = ngoService;
        }

        // ---------------- CREATE RESOURCE REQUEST ----------------
        [HttpPost("requests/resource")]
        public async Task<IActionResult> CreateResourceRequest([FromBody] CreateResourceRequestDto dto, [FromQuery] int ngoId)
        {
            await _ngoService.CreateResourceRequestAsync(ngoId, dto);
            return Ok("Resource Request created successfully");
        }

        // ---------------- CREATE VOLUNTEER REQUEST ----------------
        [HttpPost("requests/volunteer")]
        public async Task<IActionResult> CreateVolunteerRequest([FromBody] CreateVolunteerRequestDto dto, [FromQuery] int ngoId)
        {
            await _ngoService.CreateVolunteerRequestAsync(ngoId, dto);
            return Ok("Volunteer Request created successfully");
        }
        
        // ---------------- CREATE FUNDRAISER REQUEST ----------------
        [HttpPost("requests/fundraiser")]
        public async Task<IActionResult> CreateFundraiserRequest([FromBody] CreateFundraiserRequestDto dto, [FromQuery] int ngoId)
        {
            await _ngoService.CreateFundraiserRequestAsync(ngoId, dto);
            return Ok("Fundraiser Request created successfully");
        }

        // ---------------- VIEW OWN REQUESTS ----------------
        [HttpGet("{ngoId}/requests")]
        public async Task<IActionResult> ViewOwnRequests(int ngoId)
        {
            var result = await _ngoService.GetNgoRequestsAsync(ngoId);
            return Ok(result);
        }

        // ---------------- VIEW REQUEST DETAILS ----------------
        [HttpGet("requests/{requestId}/details")]
        public async Task<IActionResult> ViewRequestDetails(int requestId)
        {
            var result = await _ngoService.GetRequestDetailsAsync(requestId);
            return Ok(result);
        }
        
        // ---------------- UPDATE NGO PROFILE ----------------
        [HttpPut("{ngoId}")]
        public async Task<IActionResult> UpdateNgoDetails(int ngoId, [FromBody] UpdateNgoDto dto)
        {
            if (!ModelState.IsValid) return BadRequest(ModelState);
            await _ngoService.UpdateNgoProfileAsync(ngoId, dto);
            return Ok("NGO details updated successfully.");
        }

        // ---------------- FULFILL REQUEST ----------------
        [HttpPut("requests/{requestId}/fulfill")]
        public async Task<IActionResult> FulfillRequest(int requestId)
        {
            await _ngoService.FulfillRequestAsync(requestId);
            return Ok("Request marked as fulfilled");
        }
        
        // ---------------- CLOSE REQUEST (Explicit) ----------------
        [HttpPut("requests/{requestId}/close")]
        public async Task<IActionResult> CloseRequest(int requestId, [FromQuery] int ngoId)
        {
            await _ngoService.CloseRequestAsync(requestId, ngoId);
            return Ok("Request closed successfully");
        }
        
        // ---------------- GET DONATIONS ----------------
        [HttpGet("{ngoId}/donations")]
        public async Task<IActionResult> GetDonations(int ngoId)
        {
            var result = await _ngoService.GetNgoDonationsAsync(ngoId);
            return Ok(result);
        }
        
        // ---------------- GET FUNDRAISER DONATIONS ----------------
        [HttpGet("fundraiser-donations")]
        public async Task<IActionResult> GetFundraiserDonations([FromQuery] int ngoId)
        {
            var result = await _ngoService.GetFundraiserDonationsAsync(ngoId);
            return Ok(result);
        }
        
        // ---------------- SEND DONATION RESPONSE ----------------
        [HttpPost("donation-response")]
        public async Task<IActionResult> SendDonationResponse([FromBody] DonationResponseRequestDto dto)
        {
            await _ngoService.SendDonationResponseAsync(dto);
            return Ok("Response sent successfully");
        }
    }
}

File: backend-java/sahayaSetu/src/main/java/com/sahayaSetu/controllers/NgoController.java
package com.sahayaSetu.controllers;

import com.sahayaSetu.dtos.*;

import com.sahayaSetu.services.INgoService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;

import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/ngo")
@RequiredArgsConstructor
public class NgoController {

    private final INgoService ngoService;

    // Token verification removed for testing as per user request
    // private Long getCurrentNgoId(Authentication authentication) { ... }

    @PostMapping("/requests/resource")
    public ResponseEntity<?> createResourceRequest(@Valid @RequestBody ResourceRequestDto dto,
            @RequestParam Long ngoId) {
        ngoService.createResourceRequest(ngoId, dto);
        return ResponseEntity.ok("Resource Request created successfully");
    }

    @PostMapping("/requests/volunteer")
    public ResponseEntity<?> createVolunteerRequest(@Valid @RequestBody VolunteerRequestDto dto,
            @RequestParam Long ngoId) {
        ngoService.createVolunteerRequest(ngoId, dto);
        return ResponseEntity.ok("Volunteer Request created successfully");
    }

    @PostMapping("/requests/fundraiser")
    public ResponseEntity<?> createFundraiserRequest(@Valid @RequestBody FundraiserRequestDto dto,
            @RequestParam Long ngoId) {
        ngoService.createFundraiserRequest(ngoId, dto);
        return ResponseEntity.ok("Fundraiser Request created successfully");
    }

    @GetMapping("/{ngoId}/requests")
    public ResponseEntity<?> getNgoRequests(@PathVariable Long ngoId) {
        return ResponseEntity.ok(ngoService.getNgoRequests(ngoId));
    }

    @GetMapping("/{ngoId}/donations")
    public ResponseEntity<?> getNgoDonations(@PathVariable Long ngoId) {
        return ResponseEntity.ok(ngoService.getNgoDonations(ngoId));
    }

    @PutMapping("/{ngoId}")
    public ResponseEntity<?> updateProfile(@PathVariable Long ngoId, @RequestBody NgoUpdateDto dto) {
        ngoService.updateNgoProfile(ngoId, dto);
        return ResponseEntity.ok("Profile updated");
    }

    @GetMapping("/fundraiser-donations")
    public ResponseEntity<?> getFundraiserDonations(@RequestParam Long ngoId) {
        return ResponseEntity.ok(ngoService.getFundraiserDonations(ngoId));
    }

    @PostMapping("/donation-response")
    public ResponseEntity<?> sendDonationResponse(@RequestBody DonationResponseRequestDto dto) {
        ngoService.sendDonationResponse(dto);
        return ResponseEntity.ok("Response sent successfully");
    }

    @PutMapping("/requests/{requestId}/close")
    public ResponseEntity<?> closeRequest(@PathVariable Long requestId, @RequestParam Long ngoId) {
        ngoService.closeRequest(requestId, ngoId);
        return ResponseEntity.ok("Request closed successfully");
    }

    @GetMapping("/requests/{requestId}/details")
    public ResponseEntity<?> getRequestDetails(@PathVariable Long requestId) {
        return ResponseEntity.ok(ngoService.getRequestDetails(requestId));
    }
}
